pkg/store/copr/region_cache.go

```diff
@@ -99,195 +99,292 @@ func formatRanges(ranges *KeyRanges) zap.Field {
+func checkRangesCovered(ctx context.Context, kvRanges []tikv.KeyRange, locs []*tikv.KeyLocation) ([]bool, bool) {
+	locUsed := make([]bool, len(locs))
+	valid := true
 	locIdx := 0
-	// Track the first mismatch for better diagnostics
-	firstMismatchRangeIdx := -1
-	firstMismatchLocIdx := -1
-	firstMismatchReason := ""
-	var firstMismatchLoc *tikv.KeyLocation
-	var firstMismatchRange tikv.KeyRange
-	// Track if current range continues from previous location (partial coverage)
-	rangeContinuesFromPrevLoc := false
-	var prevLocEndKey []byte
-
-	for _, loc := range locs {
-		if loc == nil {
-			continue // Skip nil locations
-		}
-		if rangeIdx >= len(kvRanges) {
-			// All ranges processed - remaining locations are okay
+
+	for rangeIdx, r := range kvRanges {
+		// Advance to the first location that might contain r.StartKey
+		for locIdx < len(locs) {
+			loc := locs[locIdx]
+			if loc == nil {
+				locIdx++
+				continue
+			}
+			// If loc.EndKey <= r.StartKey, this loc is entirely before the range
+			if len(loc.EndKey) > 0 && compareKeyRangeBoundary(loc.EndKey, r.StartKey, false, true) <= 0 {
+				locIdx++
+				continue
+			}
 			break
 		}
-		currentRange := kvRanges[rangeIdx]
-
-		// Only validate start coverage if this is the first location for this range
-		if !rangeContinuesFromPrevLoc {
-			startCovered := false
-			if len(currentRange.StartKey) == 0 {
-				// Empty start key means beginning of key space
-				// Location must also start from beginning
-				startCovered = len(loc.StartKey) == 0
-			} else {
-				// Non-empty start key
-				startCovered = loc.Contains(currentRange.StartKey) || bytes.Equal(currentRange.StartKey, loc.StartKey)
-			}
 
-			if !startCovered && firstMismatchRangeIdx == -1 {
-				firstMismatchRangeIdx = rangeIdx
-				firstMismatchLocIdx = locIdx
-				firstMismatchReason = "location does not cover range start"
-				firstMismatchLoc = loc
-				firstMismatchRange = currentRange
-			}
-		} else {
-			// Range continues from previous location - verify no gap
-			if !bytes.Equal(prevLocEndKey, loc.StartKey) && firstMismatchRangeIdx == -1 {
-				firstMismatchRangeIdx = rangeIdx
-				firstMismatchLocIdx = locIdx
-				firstMismatchReason = "gap between locations"
-				firstMismatchLoc = loc
-				firstMismatchRange = currentRange
+		// Check if current location covers the range start
+		if locIdx >= len(locs) || !locContainsStartKey(locs[locIdx], r.StartKey) {
+			logutil.Logger(ctx).Warn("[validateLocationCoverage] range start not covered",
+				zap.Int("rangeIndex", rangeIdx),
+				keyField("rangeStart", r.StartKey),
+				keyField("rangeEnd", r.EndKey),
+				zap.Int("locIndex", locIdx),
+				zap.Int("totalLocs", len(locs)))
+			if locIdx < len(locs) && locs[locIdx] != nil {
+				logutil.Logger(ctx).Warn("[validateLocationCoverage] nearest location",
+					zap.Uint64("regionID", locs[locIdx].Region.GetID()),
+					keyField("locStart", locs[locIdx].StartKey),
+					keyField("locEnd", locs[locIdx].EndKey))
```

Comment:
Perhaps also print the previous locaction if there is one?

```suggestion
				logutil.Logger(ctx).Warn("[validateLocationCoverage] next location",
					zap.Uint64("regionID", locs[locIdx].Region.GetID()),
					keyField("locStart", locs[locIdx].StartKey),
					keyField("locEnd", locs[locIdx].EndKey))
```